lib/blockstore/file.js

```diff
@@ -0,0 +1,579 @@
+    let filerecord = null;
+    let filepath = null;
+
+    const last = await this.db.get(layout.F.encode(type));
+    if (last)
+      fileno = bio.read(last).readU32();
+
+    filepath = this.filepath(type, fileno);
+
+    const rec = await this.db.get(layout.f.encode(type, fileno));
+
+    let touch = false;
+
+    if (rec) {
+      filerecord = FileRecord.fromRaw(rec);
+    } else {
+      touch = true;
+      filerecord = new FileRecord({
+        blocks: 0,
+        used: 0,
+        length: this.maxFileLength
+      });
+    }
+
+    if (filerecord.used + length > filerecord.length) {
+      fileno += 1;
+      filepath = this.filepath(type, fileno);
+      touch = true;
+      filerecord = new FileRecord({
+        blocks: 0,
+        used: 0,
+        length: this.maxFileLength
+      });
+    }
+
+    if (touch) {
+      const fd = await fs.open(filepath, 'w');
+      await fs.close(fd);
+    }
+
+    return {fileno, filerecord, filepath};
+  }
+
+  /**
+   * This method stores block undo coin data in files.
+   * @param {Buffer} hash - The block hash
+   * @param {Buffer} data - The block data
+   * @returns {Promise}
+   */
+
+  async writeUndo(hash, data) {
+    return this._write(types.UNDO, hash, data);
+  }
+
+  /**
+   * This method stores block data in files.
+   * @param {Buffer} hash - The block hash
+   * @param {Buffer} data - The block data
+   * @returns {Promise}
+   */
+
+  async write(hash, data) {
+    return this._write(types.BLOCK, hash, data);
+  }
+
+  /**
+   * This method stores block data in files with by appending
+   * data to the last written file and updating indexes to point
+   * to the file and position.
+   * @private
+   * @param {Number} type - The type of block data
+   * @param {Buffer} hash - The block hash
+   * @param {Buffer} data - The block data
+   * @returns {Promise}
+   */
+
+  async _write(type, hash, data) {
+    if (this.writing)
+      throw new Error('Already writing.');
+
+    this.writing = true;
+
+    if (await this.db.has(layout.b.encode(type, hash))) {
+      this.writing = false;
+      return false;
+    }
+
+    let mlength = 8;
+
+    // Hash for a block is not stored with
+    // the magic prefix as it's read from the header
+    // of the block data.
+    if (type !== types.BLOCK)
+      mlength += 32;
+
+    const blength = data.length;
+    const length = data.length + mlength;
+
+    const bwm = bio.write(mlength);
+
+    bwm.writeU32(this.network.magic);
+    bwm.writeU32(blength);
+
+    if (type !== types.BLOCK)
+      bwm.writeHash(hash);
+
+    const magic = bwm.render();
+
+    const {
+      fileno,
+      filerecord,
+      filepath
+    } = await this.allocate(type, length);
+
+    const mposition = filerecord.used;
+    const bposition = filerecord.used + mlength;
+
+    const fd = await fs.open(filepath, 'r+');
+
+    const mwritten = await fs.write(fd, magic, 0, mlength, mposition);
+    const bwritten = await fs.write(fd, data, 0, blength, bposition);
```

Comment:
These two lines should be wrapped in a `try {...} finally { fs.close(fd); }` in case one of them throws.

Reply:
Added